### DISCUSSION TITLE
How to close cobbection

### DISCUSSION THREAD

**Discussion by 4dreadhead** [mitmproxy/mitmproxy#6522] [Q&A]
I am writing a module that collects tls fingerprints.

I need to close the connection between the proxy and the client after the client receives its http response. The problem is that if I keep the connection open, the tls session message remains the same, and my task is to initiate a new session every time the page is refreshed. The only thing I found is `flow.kill()`, but it kills the connection before the client receives a response from the server, so it doesn't work. And there's another problem, when the connection is still broken (apparently, by timeout), the browser tries to resume it, and in tls there are unnecessary extensions for my task, which are not present at the first connection. For example, everything I described is implemented on https://tls.browserleaks.com/tls, there every page refresh is like a new session without unexpected headers in *tls client hello* message - (41 - pre_shared_key, that not exists on first connect)

My code:
```python
import sys
import os
import json
import time
from mitmproxy import http
from mitmproxy import tls
from dotenv import load_dotenv

sys.path.append(os.getcwd())

from lib.analyzers.tls.parser import TlsParser

load_dotenv()


class Mitmproxy:
    def __init__(self):
        self.proxy_destination_port = int(os.getenv("API_PORT"))
        self.proxy_destination_host = os.getenv("API_HOST")
        self.tls_client = {}
        self.clean_timeout = 60
        self.last_clean_time = time.time()

    def request(self, flow: http.HTTPFlow):
        flow.request.scheme = "https"
        flow.request.host = self.proxy_destination_host
        flow.request.port = self.proxy_destination_port
        flow.request.headers["Connection"] = "close"
        flow.request.headers["tls"] = self.tls_client.get(flow.client_conn.peername, "{}")

    def response(self, flow: http.HTTPFlow):
        flow.response.headers["Cache-Control"] = "no-store, no-cache, must-revalidate, max-age=0"
        flow.response.headers["Pragma"] = "no-cache"

    def tls_clienthello(self, data: tls.ClientHelloData):
        current_time = time.time()
        if current_time - self.last_clean_time > self.clean_timeout:
            self.last_clean_time = current_time
            to_delete = []

            for key, value in self.tls_client.items():
                fp = json.loads(value)
                if current_time - fp[TlsParser.TIMESTAMP_KEY] > self.clean_timeout:
                    to_delete.append(key)

            for key in to_delete:
                del self.tls_client[key]

        if not self.tls_client.get(data.context.client.peername):
            self.tls_client[data.context.client.peername] = TlsParser(data.client_hello.raw_bytes()).as_str()


addons = [Mitmproxy()]
```
I run mitmproxy with command:
```
mitmdump -s lib/proxy/interceptor.py --certs $SSL_FULL --listen-port $PROXY_PORT --set tls_version_client_max=TLS1_3 --set tls_version_client_min=TLS1_3
```
Rummaged through the mitmproxy documentation, and couldn't find a tool that would allow me to accomplish my task

**Comment by mhils** [comment]
If you want to close the connection after the first response, an easy solution would be to stick to HTTP/1 and send a `Connection: close` header.
